package roomescape.dao;

import java.time.LocalDate;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
import org.springframework.stereotype.Repository;
import roomescape.domain.Reservation;
import roomescape.domain.ReservationTime;

@Repository
public class ReservationDao {

    private final JdbcTemplate jdbcTemplate;
    private final SimpleJdbcInsert simpleJdbcInsert;

    public ReservationDao(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
        simpleJdbcInsert = new SimpleJdbcInsert(jdbcTemplate)
            .withTableName("reservation")
            .usingGeneratedKeyColumns("id");
    }

    public List<Reservation> findAll() {
        return jdbcTemplate.query(
            "SELECT "
                + "r.id as reservation_id, "
                + "r.name, "
                + "r.date, "
                + "t.id as time_id, "
                + "t.start_at as time_value "
                + "FROM reservation as r "
                + "inner join reservation_time as t "
                + "on r.time_id = t.id",
            (resultSet, rowNum) ->
            {
                LocalDate date = LocalDate.parse(
                    resultSet.getString("date"),
                    DateTimeFormatter.ofPattern("yyyy-MM-dd")
                );
                return new Reservation(
                    resultSet.getLong("reservation_id"),
                    resultSet.getString("name"),
                    date,
                    new ReservationTime(
                        resultSet.getLong("time_id"),
                        LocalTime.parse(
                            resultSet.getString("time_value"),
                            DateTimeFormatter.ofPattern("HH:mm")
                        )
                    )
                );
            }
        );
    }

    public Reservation save(Reservation reservation) {
        Map<String, Object> parameters = new HashMap<>();
        parameters.put("name", reservation.getName());
        parameters.put("date", reservation.getDate());
        parameters.put("time_id", reservation.getTime().getId());
        Long id = simpleJdbcInsert.executeAndReturnKey(parameters).longValue();
        return new Reservation(id, reservation.getName(), reservation.getDate(),
            reservation.getTime());
    }

    public int deleteById(Long id) {
        return jdbcTemplate.update(
            "delete from reservation where id = ?",
            id
        );
    }
}
